src/transformers/models/sam/image_processing_sam.py

```diff
@@ -0,0 +1,658 @@
+        return int(math.ceil((overlap * (n_crops - 1) + orig_len) / n_crops))
+
+    for i_layer in range(n_layers):
+        n_crops_per_side = 2 ** (i_layer + 1)
+        overlap = int(overlap_ratio * short_side * (2 / n_crops_per_side))
+
+        crop_w = crop_len(im_w, n_crops_per_side, overlap)
+        crop_h = crop_len(im_h, n_crops_per_side, overlap)
+
+        crop_box_x0 = [int((crop_w - overlap) * i) for i in range(n_crops_per_side)]
+        crop_box_y0 = [int((crop_h - overlap) * i) for i in range(n_crops_per_side)]
+
+        # Crops in XYWH format
+        for x0, y0 in product(crop_box_x0, crop_box_y0):
+            box = [x0, y0, min(x0 + crop_w, im_w), min(y0 + crop_h, im_h)]
+            crop_boxes.append(box)
+            layer_idxs.append(i_layer + 1)
+
+    # generate cropped images
+    cropped_images = []
+    total_points_per_crop = []
+    for i, crop_box in enumerate(crop_boxes):
+        x0, y0, x1, y1 = crop_box
+        cropped_im = image[y0:y1, x0:x1, :]
+        cropped_images.append(cropped_im)
+
+        cropped_im_size = cropped_im.shape[:2]
+        points_scale = np.array(cropped_im_size)[None, ::-1]
+
+        total_points_per_crop.append(points_grid[layer_idxs[i]] * points_scale)
+
+    normalized_total_points_per_crop = []
+    # is all of this done in the processor too?
+    for points_per_crop in total_points_per_crop:
+        normalized_total_points_per_crop.append(
+            [_normalize_coordinates(target_size, point, original_size) for point in points_per_crop]
+        )
+    # normaliwation should happen inside the processor
+
+    crop_boxes = torch.tensor(crop_boxes, dtype=torch.float32)
+    points_per_crop = torch.cat(
+        [torch.tensor(p).unsqueeze(0) for p in np.array(normalized_total_points_per_crop)], dim=0
+    )
+    points_per_crop = points_per_crop.permute(1, 0, 2)
+    return crop_boxes, points_per_crop, cropped_images
+
+
+def _calculate_stability_score(masks, mask_threshold, threshold_offset):
+    """
+    Computes the stability score for a batch of masks. The stability score is the IoU between the binary masks obtained
+    by thresholding the predicted mask logits at high and low values.
+    """
+    # One mask is always contained inside the other.
+    # Save memory by preventing unnecesary cast to torch.int64
+    intersections = (masks > (mask_threshold + threshold_offset)).sum(-1, dtype=torch.int16).sum(-1, dtype=torch.int32)
+    unions = (masks > (mask_threshold - threshold_offset)).sum(-1, dtype=torch.int16).sum(-1, dtype=torch.int32)
+    return intersections / unions
+
+
+def _uncrop_boxes_xyxy(boxes, crop_box):
+    x0, y0, _, _ = crop_box
+    offset = torch.tensor([[x0, y0, x0, y0]], device=boxes.device)
+    # Check if boxes has a channel dimension
+    if len(boxes.shape) == 3:
+        offset = offset.unsqueeze(1)
+    return boxes + offset
+
+
+def _uncrop_masks(masks, crop_box: List[int], orig_h: int, orig_w: int):
+    x0, y0, x1, y1 = crop_box
```

Comment:
Very confusing here. It seems x1 and y1 are width and height? But the usual format for a box given with two points coordinates is the left/top corner and bottom/right corner. Again please be more explicit in the variable names.

Reply:
Same comment as here: https://github.com/huggingface/transformers/pull/22654#discussion_r1171043296
But let's keep all that in mind for #22840 @ArthurZucker